xla/service/while_loop_unroller.cc

```diff
@@ -755,10 +746,14 @@ absl::StatusOr<bool> WhileLoopUnroller::Run(
   // unroll. We do this ahead of time so we don't have to worry about mutating
   // the lists of computations or instructions while we iterate.
   std::vector<std::pair<HloInstruction*, WhileLoopConfig>>
-      unrollable_while_ops = GetUnrollableLoops(module, execution_threads);
+      unrollable_while_ops = GetUnrollableLoops(module, execution_threads, unroll_config_);
 
   VLOG(3) << "Number of while instructions in the module to unroll: "
           << unrollable_while_ops.size();
+  // Run pass pipeline specified by user immediately before unroll.
+  if (before_unroll_pipeline_) {
```

Comment:
Logically, this seems like a preparation step for the module before unrolling.

I think it might make more sense to execute this pipeline as part of `PrepareModuleForUnrolling` (e.g. at the end of it), as opposed to in the middle of the pass. I think it'd have the same effect as what you're trying to achieve here, but might separate the logic more cleanly (at the cost of adding another parameter to a static method, which seems OK). WDYT?

This also improves testability a little bit, in the sense that you could now call the `PrepareModuleForUnrolling` static method directly in your test to verify that preparation calls your pipeline.

Reply:
Certain passes that are proprietary to our hardware cannot be run there. They have to be run at the location I specified. Please see my discussion earlier with another reviewer about this.

Reply:
> Certain passes that are proprietary to our hardware cannot be run there. They have to be run at the location I specified. Please see my discussion earlier with another reviewer about this.

I saw this discussion, but I didn't buy that argument at face value (and I obviously don't know your passes): nothing happens between the call to `PrepareModuleForUnrolling`  and this conditional, beyond

1. finding all the while ops in the module, and
2. finding all the while ops we think we can unroll.

For your passes to cause a difference there would mean that your proprietary passes (a) insert new while ops that you *don't* want to process, or that (b) they remove/modify while ops that you wanted to (somehow) aggregate in these vectors. (b) doesn't seem like something one'd want to do at all, and (a) seems like something that could be done by inserting the while ops after the `WhileLoopUnroller` pass has run, or before by inserting loops that can not be unrolled.

For any other use case, it seems like it should make no practical difference whether this pass pipeline is called here or as the last step in `PrepareModuleForUnrolling`. However, in terms of code organization, it makes much more sense to me for this pipeline to be called there.

Am I missing something? Are you truly sure that putting it here should make a difference compared to putting this at the end of `PrepareModuleForUnrolling`?

Reply:
Yes, I am sure. For some of our proprietary passes if you run them before WhileLoopUnroller then while loop unroller will say the body cannot be unrolled. This is due to some of the transformations we make. The body can still be unrolled so thats why we want to run the passes directly before we call the unroll function.

Reply:
> Yes, I am sure. For some of our proprietary passes if you run them before WhileLoopUnroller then while loop unroller will say the body cannot be unrolled. This is due to some of the transformations we make. The body can still be unrolled so thats why we want to run the passes directly before we call the unroll function.

Just to make sure we didn't misunderstand each other, note that I wasn't suggesting actually running your pass pipeline before the`WhileLoopUnroller`, but at the end of `PrepareModuleForUnrolling` (which is firmly inside the `WhileLoopUnroller` pass, and called just a few lines above).

Reply:
Sorry I meant PrepareModuleForUnrolling. If you run it in PrepareModuleForUnrolling, GetUnrollableLoops will say the loops cannot be unrolled. This is because some of our transformations break those conditions. Hence we run the transformation immediately before the unroll call itself. 
https://github.com/openxla/xla/blob/d5d9a751e03fc147c5c64160feb7602f6348eb6c/xla/service/while_loop_unroller.cc#L749

Reply:
I'm gonna suggest that you look at why the PrepareModuleForUnrolling is not sufficient to make the loops in question (the ones that got transformed by your passes) unrollable.

Reply:
Without explaining too much, these proprietary passes insert custom-calls inside the body of the while loop. Why does GetUnrollableLoops fail if there are custom-calls in the body of the body region? I can try and debug that, but if that is by design I do not understand why providing power users (hardware vendors) the option to control exactly where they want to run some passes for unrolling. This change is mainly for power users of XLA, so I do not understand the pushback.

Reply:
The reason is that this particular design choice might be an artifact of a more serious invariant that is breaking here and I overlooked earlier. I think the fact that GetUnrollableLoops fails if there are custom-calls added to the while body, by itself is a red flag. These custom-calls might be in fact the reason that make the loop unrollable (e.g, they have side effect and are inside the condition which we end up removing after unrolling). I suggest that you look at the IsLoopUnrollable function and see if those custom-calls are violating any conditions there. I believe after this investigation we are able to better decide how to move forward.

Reply:
OK I can investigate this. The loop is definitely unrollable with the custom-calls, as the custom-calls are no ops. They are our proprietary custom-calls but have no real function.

Reply:
I have reduced a very small test case that can trigger the unroller to refuse to unroll.

Reply:
Found the problem. custom-calls at start and end of the loop break 
  std::optional<int64_t> indvar_tuple_idx =
      GetLoopInductionVarTupleIdx(while_op);
  if (!indvar_tuple_idx.has_value()) {
    LOG(INFO) << "Break on line 582";
    return std::nullopt;
  }

Reply:
This is the problem @fhoushmand @bchetioui 
```
I0000 00:00:1723832752.201489   52013 while_loop_analysis.cc:293] While body's root is not a tuple instruction: %custom-call.2 = (s32[]{:T(128)}, s32[3]{0}) custom-call((s32[]{:T(128)}, s32[3]{0}) %tuple), custom_call_target="CustomCallEnd"
```

Can I add an optional config to GetLoopInductionVarTupleIdx saying allow_custom_calls=True? Now you see why I wanted to run the pass pipeline immediately before unroll time as some passes break the induction var pattern match. I still think my original approach is preferred.

Reply:
Reproducible test case 
```
TEST_F(WhileLoopUnrollerTest, SimpleLoopWithCustomCall){
  std::string hlo_string = R"(
  HloModule SimpleLoop
  SimpleLoop.body {
    loop_var.1 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
    custom-call.1 = (s32[]{:T(128)}, s32[3]{0}) custom-call(loop_var.1), custom_call_target="CustomCallStart"
    get-tuple-element.1 = s32[]{:T(128)} get-tuple-element(custom-call.1), index=0
    constant.1 = s32[]{:T(128)} constant(1)
    idx = s32[]{:T(128)} add(get-tuple-element.1, constant.1)
    get-tuple-element.2 = s32[3]{0} get-tuple-element(custom-call.1), index=1
    output = s32[3]{0} add(get-tuple-element.2, get-tuple-element.2)
    tuple = (s32[]{:T(128)}, s32[3]{0}) tuple(idx, output)
    ROOT custom-call.2 = (s32[]{:T(128)}, s32[3]{0}) custom-call(tuple), custom_call_target="CustomCallEnd"
  }
  SimpleLoop.condition {
    loop_var.2 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
    get-tuple-element.3 = s32[] get-tuple-element(loop_var.2), index=0
    constant.2 = s32[]{:T(128)} constant(5)
    ROOT less-than = pred[] compare(get-tuple-element.3, constant.2), direction=LT
  }
  ENTRY SimpleLoop {
    constant.3 = s32[]{:T(128)} constant(0)
    constant.4 = s32[3]{0} constant({0, 1, 2})
    tuple.1 = (s32[]{:T(128)}, s32[3]{0}) tuple(constant.3, constant.4)
    ROOT while = (s32[]{:T(128)}, s32[3]{0}) while(tuple.1), condition=
      SimpleLoop.condition, body=SimpleLoop.body
  }
  )";
  auto m = ParseAndReturnVerifiedModule(hlo_string).value();
  UnrollConfig config;
  EXPECT_TRUE(WhileLoopUnroller(/*unroll_factor=*/-1, /*wrap_in_trivial_loop=*/false, config).Run(m.get()).value());


}
```

Reply:
Thanks @ptoulme-aws for looking into this.

> Can I add an optional config to GetLoopInductionVarTupleIdx saying allow_custom_calls=True? 
Caller should not care about what kind of instructions are in the while body. If we can get the induction var statically at compile time, that function should be able to handle it without needing to pass additional arguments. The reason that this was not handled in the function is this pattern was never observed. You can add it though.

> Now you see why I wanted to run the pass pipeline immediately before unroll time as some passes break the induction var pattern match. I still think my original approach is preferred.

I'm gonna have to disagree with you on that. Perhaps, your approach might be easier, but it would be dangerous. Assume that the pipeline is given a pass that can change the trip count of the loops. It's easy to see that doing surgery in the middle of the pass would break invariants that the pass depends on.

I like to see each compiler pass as an atomic unit that operates on a module that satisfies a pre-condition, and produces a module satisfying a post-condition. Given this, one must be cautious when breaking atomicity.

Reply:
Thank you. I will add the pattern and remove the interior pass pipeline I added and add tests for new patterns.